### DISCUSSION TITLE
A stable ABI (at least for GSP FW)

### DISCUSSION THREAD

**Discussion by edisionnano** [NVIDIA/open-gpu-kernel-modules#372] [Ideas]
During the Linux Plumbers Conference 2022 the Nouveau devs held a presentation which mostly concerned the future of the project after NVIDIA released this open source kernel drivers which reignited the interest in Nouveau project.
One of their most important concerns is the lack of a stable ABI. What this means is that NVIDIA drivers are version locked. For example, the latest kernel module is version 515.49.15 and only userspace libraries and GSP firmware version 515.49.15 is compatible with it. Nouveau is an upstream driver and you can see why this is an issue.
I'd like to request a better design with a stable ABI
https://youtube.com/watch?v=KkOdMwZRpYY at 08:49:20

**Comment by mtijanic** [comment]
Hey there. Yes, lack of stable ABI is a known problem with getting an upstream GSP driver - whether Nouveau or a new one. We have been thinking about this for a long time and the infrastructural groundwork is mostly in place. 
If you look at some of the APIs for talking to GSP, such as [this file](https://github.com/NVIDIA/open-gpu-kernel-modules/blob/main/src/common/sdk/nvidia/inc/ctrl/ctrl2080/ctrl2080internal.h) you'll see that the C code is generated from a higher level Interface Definition Language. Longer term, we will generate API+ABI-stable messages - this could look similar to e.g. protobuf messages with optional fields.

The major issue here is that once we start supporting a stable, versioned ABI, we need to carry all the additional baggage for a long time. And GSP development is still moving very rapidly - just this week we've updated at least 3 RPCs (that I know of). If this was all versioned we'd accumulate so much tech debt before we ever shipped a production ready geforce driver.

Bottom line is, we are aware of the issues this causes for third party kernel drivers, and we do have plans on how to resolve it, we just can't execute on those plans yet or everyone will be paying the price for a long time to come. Fortunately, this shouldn't preclude the _development_ of such drivers - yes there will be a bit of throwaway work, but getting the r515 GSP image booting is 98% of the work, and then switching it use a versioned interface later should not be much of an issue.

**Reply by edisionnano** [reply]
Hi, thanks for the quick response. What happens to other firmware (eg. PMU), are they version locked too?

**Reply by mtijanic** [reply]
They are, but with GSP it doesn't matter. This touches on the other issues Dave raised - the method to figure out which firmware to load and how is very arcane and fragile. Sometimes the load order varies from chip to chip. But, this is now all handled by GSP, and all a kernel driver needs to do is load the GSP (which, admittedly, is a lot more complex than you'd expect).

This is also why the gsp.bin is 30+ megs. The proprietary `nvidia.ko` embeds over a thousand different firmwares into the [bindata infrastructure](https://github.com/NVIDIA/open-gpu-kernel-modules/blob/main/src/nvidia/inc/kernel/core/bin_data.h), and then it knows if/when to fetch and load them onto the different microcontrollers. All of this has been moved into GSP, both the loading logic and the thousand smaller firmware files[0]. Since GSP is version-locked with this firmware, and since kernel no longer talks to PMU directly, the kernel code shouldn't care about this interface at all. If the GSP API/ABI are stable, nouveau shouldn't care about the PMU API/ABI.

The gsp.bin size problem is also known, but there are confounding factors that make it not trivial to resolve, so I can't promise anything there yet except that we are working on it.

Few related discussions if you're curious:
- https://github.com/NVIDIA/open-gpu-kernel-modules/issues/286
- https://github.com/NVIDIA/open-gpu-kernel-modules/issues/62#issuecomment-1125077998 


[0]: I know "they moved 30mb of code from kernel into firmware!" makes for a good story, but we never had 30mb of kernel code to begin with. Most of it was always just embedded firmware blobs.

**Reply by throwaway1037** [reply]
The issue is not specifically that the 30+mb of blobs was in one place and was moved to another, the issue is that NVIDIA is releasing blobbed drivers under the false label of "open" and then closing, locking, and ignoring all the issues which ask them to release the blobs under a free software license, as free/libre software.
It's now been over 4 months since the initial release of this repo, and the drivers are still unusable without proprietary software and firmware.

This feels like a spit in the face to the very people who want to run in freedom the GPUs they have already purchased, a subset of whom also want to cooperate and help develop these drivers and firmware, with no financial incentive whatsoever.
This whole ordeal looks more and more like a purely strategic business move because NVIDIA wants to get good publicity and be competitive in data centers, not a sincere attempt to treat the users better.

**Reply by mtijanic** [reply]
> closing, locking, ignoring all the issues which ask them to release the blobs under a free software license, as free/libre software.

We have already repeatedly said that there are currently no plans to release the source code of the firmware blobs. You do have active discussion threads where you are free to discuss the implications of this decision; we have no desire to censor anyone (as long as it's kept civil). I'm not sure what else you expect of us (the moderators) to do? If there are any facts that we can clarify, please ask explicitly and we'll do the best we can. Otherwise we stay silent and "ignore" because we have nothing useful to add to the discussions.

**Reply by throwaway1037** [reply]
I understand there's only so much NVIDIA engineers and staff can do, but do the bosses who make the decisions to release this code under a free software license know about this situation and realise the issues people have raised?
At worst, the bosses are ignoring us, and NVIDIA engineers and staff who communicate the lack of decisions via GitHub can't do much to help, and at best the bosses haven't been notified yet.

**Reply by edisionnano** [reply]
I don't think the GSP not being open matters much. Ofc it would be nice if it was but it doesn't matter much. AMD also has blob fw. What matters is that this FW is distributed under a license that permits redistribution.

**Reply by throwaway1037** [reply]
It is a mistake to treat software respecting our freedom as a mere nice-to-have feature instead of a moral imperative, yet this is exactly what the rhetoric of "open source" invites us to do.
See https://gnu.org/philosophy/open-source-misses-the-point.html

What matters is our freedom, and all free software licenses permit redistribution, which also meets your criteria.
The GSP firmware being free/libre matters a lot, because without it, these drivers are unusable in the free world.

AMD publishing firmware blobs is no excuse for anyone else to do so.
Two wrongs don't make a right.

**Reply by mtijanic** [reply]
> do the bosses who make the decisions to release this code under a free software license know about this situation and realise the issues people have raised?

I can't really say if a particular "boss" is aware of a particular issue, but generally the consequences of not publishing the firmware sources (and signing keys) are well known. This forum is monitored by people who aggregate user feedback, so you have definitely been heard. I'm afraid we can't know if it made it into the "Top N" executive summary or if it was pushed out by other user feedback that was prioritized higher.

Either way, the end result is the same: There are currently no plans to publish the firmware sources. We are aware that as a consequence NVIDIA GPUs are not usable by projects such as linux-libre. We have no plans to change this.

Annoying as it may be, there is not really much more you can do about this. You are free to preach your philosophy and try to build a critical mass of users for whom this is a showstopper issue; it could work. However I have to ask you to refrain from derailing unrelated discussions with these topics and stick to the several already open threads.

I don't think there's anything else for us to say on this subject for the time being. Thank you.

**Reply by throwaway1037** [reply]
@mtijanic thank you for talking with me. I understand you're doing all you can, I just wish NVIDIA would take action. Many of these technical issues stem from the lack of freedom, and if all the components were libre, people wouldn't have to be so reliant on asking tons of questions and speculating. We want to cooperate with you, NVIDIA.

**Reply by throwaway1037** [reply]
It's possible NVIDIA wrote those blobs in binary, maybe if they were masochists, but it's much more likely they wrote them in a human-readable programming language, such as C or one of the various assembly languages.
Disassembling precompiled binaries is a form of reverse-engineering and is not source code.
Source code contains things like macros, comments, functions, and many other constructs designed to make programs human-readable.

@mtijanic please can you confirm for @ValZapod that these blobs do have source code?
I don't think he will believe me unless he hears it straight from the people working on them, and he keeps bringing it up.

**Comment by edisionnano** [comment]
@mtijanic Nouveau aims to support all Nvidia gpus (particularly Kepler and newer). From what I understand GSP fw isn't driver code moved to fw but rather all firmwares merged into a single package. While some firmwares are provided by nvidia over at linux firmware, pmu is notably missing and it does indeed matter for reclocking maxwell and pascal using Nouveau so along with stable GSP I really hope we'll see the other firmwares published too

**Reply by throwaway1037** [reply]
Not just published, but published as free/libre software, with the ability for users to sign their own firmware and use their own GPUs in freedom.
More blobs is worse, as it's one more component that has to be replaced with free software, and it sends the message that blobs are okay in Linux, which makes even more work for Linux-libre to remove them.

**Reply by mtijanic** [reply]
> From what I understand GSP fw isn't driver code moved to fw but rather all firmwares merged into a single package.

It's both of those. There is code that previously executed in kernel mode that now executes on the GSP microcontroller instead. There is also code that executes on other microcontrollers (e.g. PMU) that was previously embedded into `nvidia.ko` as a blob, but is now embedded into `gsp.bin` as a blob.

> pmu is notably missing and it does indeed matter for reclocking maxwell and pascal using Nouveau

This is a known issue, but unfortunately I don't have any news to share here, sorry.

**Reply by edisionnano** [reply]
I see thanks, that is at least a response. As a Pascal GPU owner PMU firmware is of utmost importance to me since these cards are capable and will be for a long time. Unrelated but is there any discussion or issue with everything that is missing from the Open Kernel Modules compared to the prop Linux and Windows driver?

**Reply by mtijanic** [reply]
> Unrelated but is there any discussion or issue with everything that is missing from the Open Kernel Modules compared to the prop Linux and Windows driver?

AFAIK there is no consolidated place for this. The [driver readme](https://us.download.nvidia.com/XFree86/Linux-x86_64/515.57/README/kernel_open.html) states:
> However, in the current release, some display and graphics features (notably: G-SYNC, Quadro Sync, SLI, Stereo, rotation in X11, and YUV 4:2:0 on Turing), as well as power management, and NVIDIA virtual GPU (vGPU), are not yet supported. These features will be added in upcoming driver releases.

There's other things not listed there, like various Notebook-specific features and Suspend/Resume (I guess included in 'power management' above). And then there's plenty of minor things that would be too numerous to mention - e.g. I believe the GPU logo illumination control is not supported.

Hopefully the list will shrink considerably over the next few (major) driver releases.

**Reply by throwaway1037** [reply]
Do you know why these features were omitted in the initial release?

**Reply by mtijanic** [reply]
> Do you know why these features were omitted in the initial release?

The removed code is currently not functional when GSP is in use. It is present as dead, non-functional code in the proprietary driver running in GSP mode. It was removed from the published source because:
1. It would just be confusing to everyone to publish useless dead code.
2. Every line of code we publish has to be read and approved. No point in doing that for stuff that doesn't work and will be rewritten anyway.

We are bringing up features with GSP as we go along. For example, the [initial datacenter-only closed source release of the GSP model](https://us.download.nvidia.com/XFree86/Linux-x86_64/470.42.01/README/gsp.html) was mid 2021 with 470.xx drivers and it did not have display capabilities at all.

**Reply by edisionnano** [reply]
Interesting, so lower CPU usage too. I believe a discussion/issue should be created here with the differences between prop and open

**Reply by throwaway1037** [reply]
This doesn't seem to add up.
Why partially liberate a crippled driver with missing functionality that depends on a 30+mb blob instead of the previous driver that had all this functionality, supported older generations, and didn't need to use the GSP?
This seems to corroborate the theory that NVIDIA moved all the interesting code into the firmware blob, refused to liberate it, and then released "open" modules as a way of getting good publicity and being competitive in data centers, whilst not providing much help to Nouveau.

@mtijanic please tell me I'm wrong.
I remember in another thread you said that the GSP blob is an aggergation of many smaller blobs that were previously embedded in other files, but if this is the case, then why also remove all the now-dead code that would be useful to Nouveau and previous gens?
NVIDIA certainly has the resources to approve that code for publication.

**Reply by aaronp24** [reply]
@throwaway1037, edisionnano opened this thread to discuss a stable API for the GSP firmware. Can you please refrain from derailing this and other threads? Milos already asked you nicely once.

**Reply by throwaway1037** [reply]
@aaronp24 perhaps I was being unclear earlier.
My point is that the issue of the blobs being proprietary is the root cause of many technical issues, so if the blobs are released as free/libre software, those technical issues would become fixable.
At the moment, we are reliant on NVIDIA to fix these issues like ABI stability for us, but if they were free/libre software, we could fix them ourselves.

Therefore, the issue of freedom is relevant to every technical issue on this tracker, and not having freedom is actively making fixing problems a nightmare.
Currently, people have to report an issue and then wait on NVIDIA to fix it, because the blobs are proprietary and half of the non-GSP code is unpublished, so we don't have a copy of that source code to work on fixing.

If you also take into account that the previous non-GSP driver was superior to this driver in terms of GPU support, and had many more fully-implemented features, then it becomes obvious to ask why this half-finished, blob-dependent driver was partially released as free/libre software (blobs aside) when the previous driver would have been much more useful to the community.

Those are my thoughts: I *am* discussing the relevant technical issue, but I recognise that the root cause is a lack of freedom, and that many of these technical issues wouldn't even exist if we had a free/libre copy of the previous, non-GSP driver.

I hope this makes sense, please ask me to clarify if I was unclear.
